### ISSUE TITLE
GHC 9.6 checklist

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/th-desugar#170]
This issue serves as a reminder to do certain things before we ship a new major release that supports GHC 9.6:

* [x] Support [`type data`](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9408)
* [x] Hackage release

**Comment by RyanGlScott** [comment]
The [`type data`-in-TH patch](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/9408) added support for `type data` declarations with the following `Dec` constructor:

```hs
data Dec
  = ...
  | TypeDataD Name [TyVarBndr ()]
          (Maybe Kind)            -- Kind signature (allowed only for GADTs)
          [Con]                   -- ^ @{ type data T x = A x | B (T x) }@
  | ...
```

We now have three different `Dec` data constructors for data type declarations: `DataD`, `NewtypeD`, and `TypeDataD`. The existing precedent in `th-desugar` is to combine these different data constructors into a single form:

```hs
data DDec
  = ...
  | DDataD NewOrData DCxt Name [DTyVarBndrUnit] (Maybe DKind) [DCon] [DDerivClause]
  | ...
```

Where `NewOrData` distinguishes between newtypes and other data types:

```hs
data NewOrData = Newtype
               | Data
```

I think we should continue this precedent and meld `TypeDataD` into `DDataD`, extending `NewOrData` in the process. I'm thinking something like this:

```hs
data NewOrData = Newtype
               | Data Bool -- 'True' for @type data@, 'False' otherwise
                           -- Alternatively, we could use some type isomorphic to Bool
```

Or even this:

```hs
data NewOrData = Newtype
               | Data
               | TypeData
```

Regardless of how we represent this, we could make sure to carefully document the invariants that `DDataD` upholds vis-à-vis `NewOrData`. Specifically:

* If a `DDataD` represents a newtype declaration, then it has exactly one `DCon`.
* If a `DDataD` represents a `type data` declaration, then it will not have any `DDerivClause`s.

**Comment by RyanGlScott** [comment]
Re-opening to track uploading `th-desugar-1.15` to Hackage.

**Comment by RyanGlScott** [comment]
Released to Hackage as [`th-desugar-1.15`](https://hackage.haskell.org/package/th-desugar-1.15).
